src/shared/nsflags.c

```diff
@@ -0,0 +1,126 @@
+/***
+  This file is part of systemd.
+
+  Copyright 2016 Lennart Poettering
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <sched.h>
+
+#include "alloc-util.h"
+#include "extract-word.h"
+#include "nsflags.h"
+#include "seccomp-util.h"
+#include "string-util.h"
+
+const struct namespace_flag_map namespace_flag_map[] = {
+        { CLONE_NEWCGROUP, "cgroup" },
+        { CLONE_NEWIPC,    "ipc"    },
+        { CLONE_NEWNET,    "net"    },
+        /* So, the mount namespace flag is called CLONE_NEWNS for historical reasons. Let's expose it here under a more
+         * explanatory name: "mnt". This is in-line with how the kernel exposes namespaces in /proc/$PID/ns. */
+        { CLONE_NEWNS,     "mnt"    },
+        { CLONE_NEWPID,    "pid"    },
+        { CLONE_NEWUSER,   "user"   },
+        { CLONE_NEWUTS,    "uts"    },
+        {}
+};
+
+const char* namespace_flag_to_string(unsigned long flag) {
+        unsigned i;
+
+        flag &= NAMESPACE_FLAGS_ALL;
+
+        for (i = 0; namespace_flag_map[i].name; i++)
```

Comment:
i did that initially, but the array is exported, since the seccomp_restrict_namespaces() call from seccomp-util.c needs it. And for exported arrays the size cannot be determined via ELEMENTSOF() that easily, unless it is either specified explicitly in the header file (which I didn't like since it might very well get out of date). The other option is to export the size as a second exported (const) variable, but that didn't sound pretty either to me. Hence I opted for a "null" terminated list here, as this allows us to export this without specifiying the size in the header, and doesn't require us to add a second exported variable. And given that all algorithms using this iterate linearly through the list anyway it doesn't really matter anyway if they know the precise size in advance or if they just look for the final NULL...

(of course, another option would have been to move seccomp_restrict_namepsaces() into nsflags.c, but that's nasty due to the fact that seccomp-util.c is currently conditionally compiled depending on ehwther the libseccomp dep is on, and we'd then have to add something similar to nsflags.c which i really didn't like)